internal/rust/semantic/src/types.rs

```diff
@@ -0,0 +1,666 @@
+            MonoType::Time => f.write_str("time"),
+            MonoType::Rgx => f.write_str("rgx"),
+            MonoType::Var(var) => var.fmt(f),
+            MonoType::Arr(arr) => arr.fmt(f),
+            MonoType::Rec(obj) => obj.fmt(f),
+            MonoType::Fun(fun) => fun.fmt(f),
+        }
+    }
+}
+
+// Tvar stands for type variable.
+// A type variable holds an unknown type.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct Tvar(i64);
+
+impl fmt::Display for Tvar {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "t{}", self.0)
+    }
+}
+
+// Array is a homogeneous list type
+#[derive(Debug, Clone, PartialEq)]
+pub struct Array(MonoType);
+
+impl fmt::Display for Array {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "[{}]", self.0)
+    }
+}
+
+// Row is an extensible record type.
+//
+// A row is either Nil meaning it has no properties,
+// or it is an extension of a row.
+//
+// A row may extend what is referred to as a row
+// variable. A row variable is a type variable that
+// represents an unknown record type.
+//
+#[derive(Debug, Clone)]
+pub enum Row {
+    Nil,
+    Var(Tvar),
+    Extension { head: Property, tail: Box<Row> },
+}
+
+impl fmt::Display for Row {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Row::Nil => f.write_str("{}"),
+            Row::Var(tv) => tv.fmt(f),
+            Row::Extension { head: h, tail: t } => write!(f, "{{{} | {}}}", h, t),
+        }
+    }
+}
+
+impl cmp::PartialEq for Row {
+    fn eq(&self, other: &Self) -> bool {
+        self.flatten() == other.flatten()
+    }
+}
+
+impl Row {
+    // Records are implemented as a sequence or list of extenstions.
+    // This function extends a record by adding a new property to the
+    // head of the list.
+    fn extend(self, head: Property) -> Self {
+        Row::Extension {
+            head: head,
+            tail: Box::new(self),
+        }
+    }
+
+    // Flatten a record type into a hashmap of property names and types
+    fn flatten(&self) -> (HashMap<String, &MonoType>, Option<Tvar>) {
+        match self {
+            Row::Nil => (HashMap::new(), None),
+            Row::Var(tv) => (HashMap::new(), Some(*tv)),
+            Row::Extension { head: h, tail: t } => {
+                let (mut r, tv) = t.flatten();
+                r.insert(h.k.clone(), &h.v);
+                (r, tv)
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Property {
+    pub k: String,
+    pub v: MonoType,
+}
+
+impl fmt::Display for Property {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}:{}", self.k, self.v)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Function {
+    pub req: HashMap<String, MonoType>,
+    pub opt: HashMap<String, MonoType>,
+    pub pipe: Option<Property>,
+    pub retn: Option<MonoType>,
```

Comment:
Should the return type be optional? For now I would like to simplify and require a return type for functions. The Flux SPEC says that functions must have a return type so I think we should follow that for now. We can investigate allowing `void` return types later.

Reply:
My bad. I was not aware that was specified in the spec. But I like that much much much better.